### PR TITLE
executor: add builtin aggregate function `json_arrayagg`

### LINKED ISSUE #7623: Call For Participation: finish the unsupported aggregate functions
All issues are open and welcomed to contributors. You can join **#sig-exec** on [tidb community slack](https://pingcap.com/tidbslack/) to discuss and get help from someone.

## Description

At present, the aggregate functions that TiDB supports are:
- `AVG()`
- `BIT_AND()`
- `BIT_OR()`
- `BIT_XOR()`
- `COUNT()`
- `COUNT(DISTINCT)`
- `GROUP_CONCAT()`
- `MAX()`
- `MIN()`
- `SUM()`

In order to be compatible with MySQL and Improve the usability of TiDB, the following functions should also be supported in TiDB:

| Name               | Description                               |
|--------------------|-------------------------------------------|
| `STD()`            | Alias for `STDDEV_POP` function  |
| `STDDEV()`         | Alias for `STDDEV_POP` function  |
| `STDDEV_POP()`     | Return the population standard deviation  |
| `STDDEV_SAMP()`    | Return the sample standard deviation      |
| `VARIANCE()`       | Return the population standard variance   |
| `VAR_POP()`        | Return the population standard variance   |
| `VAR_SAMP()`       | Return the sample variance                |
| `JSON_ARRAYAGG()`  | Return result set as a single JSON array  |
| `JSON_OBJECTAGG()` | Return result set as a single JSON object |

An implementation checklist:

- [x] `STD()`/`STDDEV()`/`STDDEV_POP()` @Win-Man  #19195
- [x] `STDDEV_SAMP()`  @Win-Man #19810
- [x] `VARIANCE()`/`VAR_POP()` @githubFZX #14101 
- [x] `VAR_SAMP()` @Win-Man #19810
- [ ] `JSON_ARRAYAGG()`  https://github.com/pingcap/tidb/pull/19957
- [x] `JSON_OBJECTAGG()` @hg2990656 #11154 

## Difficulty

* Medium

## Score 

* 3000

## Mentor(s)

* @qw4990

## Recommended Skills

* SQL Optimization
* Golang Profiling
* Code Refactoring

### LINKED ISSUE #19795: Call For Participation: implement aggregate function JSON_ARRAYAGG
All issues are open and welcomed to contributors. You can join **#sig-exec** on [tidb community slack](https://pingcap.com/tidbslack/) to discuss and get help from someone.

## Description

At present, the aggregate function `JSON_ARRARYAGG` is not supported by TiDB.

This is a part job of https://github.com/pingcap/tidb/issues/7623, please help us implement it if you are interested in it.

## Difficulty

* Medium

## Score 

* 900

## Mentor(s)

* @qw4990

## Recommended Skills

* SQL Optimization
* Golang Profiling
* Code Refactoring

### PR DESCRIPTION
**Pull request by arthuryangcs**
### What problem does this PR solve? 

* closes #7623
* closes #19795

### What is changed and how it works?

add an json_arrayagg function

### Check List 

* Make sure the replace instruction is changed back to the required instruction and the version is the latest.

### Tests 

``` golang
	tk.MustExec(`CREATE TABLE t (
		a int(11),
		b varchar(100),
		c decimal(3,2),
		d json,
		e date,
		f time,
		g datetime DEFAULT '2012-01-01',
		h timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
		i char(36),
		j text(50));`)

	tk.MustExec(`insert into t values(1, 'ab', 5.5, '{"id": 1}', '2020-01-10', '11:12:13', '2020-01-11', '0000-00-00 00:00:00', 'first', 'json_arrayagg_test');`)

	result := tk.MustQuery("select a, json_arrayagg(b) from t group by a order by a;")
	result.Check(testkit.Rows(`1 ["ab"]`))
	result = tk.MustQuery("select b, json_arrayagg(c) from t group by b order by b;")
	result.Check(testkit.Rows(`ab [5.5]`))
	result = tk.MustQuery("select e, json_arrayagg(f) from t group by e order by e;")
	result.Check(testkit.Rows(`2020-01-10 ["11:12:13"]`))
	result = tk.MustQuery("select f, json_arrayagg(g) from t group by f order by f;")
	result.Check(testkit.Rows(`11:12:13 ["2020-01-11 00:00:00"]`))
	result = tk.MustQuery("select g, json_arrayagg(h) from t group by g order by g;")
	result.Check(testkit.Rows(`2020-01-11 00:00:00 ["0000-00-00 00:00:00"]`))
	result = tk.MustQuery("select h, json_arrayagg(i) from t group by h order by h;")
	result.Check(testkit.Rows(`0000-00-00 00:00:00 ["first"]`))
	result = tk.MustQuery("select i, json_arrayagg(j) from t group by i order by i;")
	result.Check(testkit.Rows(`first ["json_arrayagg_test"]`))
	result = tk.MustQuery("select json_arrayagg(23) from t group by a order by a;")
	result.Check(testkit.Rows(`[23]`))
	result = tk.MustQuery("select json_arrayagg(null) from t group by a order by a;")
	result.Check(testkit.Rows(`[null]`))
```

### Code changes

- Add executor/aggfuncs/func_json_arrayagg.go
- Add executor/aggfuncs/func_json_arrayagg_test.go
- Delete all `srcChk.AppendDatum(0, &types.Datum{})` and the end of Chunk

### Related changes

 - Need to add the definition of json_objectagg in pingcap/parser 

### Release Note

```release-note
Add builtin aggregate function `json_arrayagg`
```

### CODE REVIEW

**Comment by kennytm** [comment]
depends on pingcap/parser#1020

**Comment by arthuryangcs** [comment]
@qw4990 @wshwsh12 @winoros Please review the code.

**Comment by arthuryangcs** [comment]
@qw4990 @wshwsh12 @winoros Please.

**Comment by jackysp** [comment]
Sorry for not replying for so long. Do you have time to resolve the conflict? @arthuryangcs

**Comment by jackysp** [comment]
Very good to see you back. @arthuryangcs 
Please fix CI, thanks!

**Comment by jackysp** [comment]
PTAL @qw4990 @XuHuaiyu @winoros

**Comment by winoros** [comment]
@arthuryangcs Hi, have you joined our community slack channel?

**Comment by arthuryangcs** [comment]
> @arthuryangcs Hi, have you joined our community slack channel?

yes.

**Comment by winoros** [comment]
Would you please send a direct message to Cui Yiding in the channel?
Sorry for your inconvenience. :joy:

**Comment by zz-jason** [comment]
To ensure the product quality of release 4.0, we need to add enough tests to ensure that this function works as expected in any situation before cherry-picking it to release-4.0.

**Comment by jackysp** [comment]
Whether it is merged to 4.0 may need to be discussed, and I also have doubts about it. The review on the master can continue.

**Comment by zz-jason** [comment]
got it. how about removing the `needs-cherry-pick-4.0` for now, adding it back after the decision is made?

**Comment by jackysp** [comment]
I am afraid of when in case it is merged, and not cherry-pick in time, then it needs to be done manually again. We can control the pr generated by the bot to decide whether to merge into release 4.0.

File: executor/aggfuncs/func_json_arrayagg.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+func (e *jsonArrayagg) AllocPartialResult() (pr PartialResult, memDelta int64) {
+	p := partialResult4JsonArrayagg{}
+	p.entries = make([]interface{}, 0)
+	return PartialResult(&p), 0
+}
+
+func (e *jsonArrayagg) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4JsonArrayagg)(pr)
+	p.entries = make([]interface{}, 0)
```
Comment (qw4990) [inline executor/aggfuncs/func_json_arrayagg.go:44]:
```suggestion
	p.entries = p.entries[:0]
```

File: expression/integration_test.go [hunk 2]
```diff
@@ -3904,6 +3904,46 @@ func (s *testIntegrationSuite) TestAggregationBuiltinGroupConcat(c *C) {
 	tk.MustQuery("select * from d").Check(testkit.Rows("hello,h"))
 }
 
+func (s *testIntegrationSuite) TestAggregationBuiltinJSONArrayagg(c *C) {
```
Comment (qw4990) [inline expression/integration_test.go:3907]: Please add more tests about `json-arrayagg distinct`.

Comment (arthuryangcs) [reply expression/integration_test.go:3907]: json_arrayagg does not support distinct in MySQL, maybe I should delete its implementation.

**Comment by jackysp** [comment]
I think you forget to delete `TestAggregationBuiltinJSONArrayaggDistinct`. @arthuryangcs

**Comment by arthuryangcs** [comment]
> I think you forget to delete `TestAggregationBuiltinJSONArrayaggDistinct`. @arthuryangcs

Done.

**Comment by jackysp** [comment]
PTAL @qw4990 @XuHuaiyu @winoros

**Comment by arthuryangcs** [comment]
PTAL @qw4990 @XuHuaiyu @winoros

**Review by jackysp — APPROVED** [review]
LGTM

File: executor/aggfuncs/func_json_arrayagg.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+type partialResult4JsonArrayagg struct {
+	entries []interface{}
+}
+
+func (e *jsonArrayagg) AllocPartialResult() (pr PartialResult, memDelta int64) {
+	p := partialResult4JsonArrayagg{}
+	p.entries = make([]interface{}, 0)
+	return PartialResult(&p), 0
```
Comment (blacktear23) [inline executor/aggfuncs/func_json_arrayagg.go:39]: Please calculate the `memDelta` instead of return 0.

File: executor/aggfuncs/builder.go [hunk 4]
```diff
@@ -611,6 +613,22 @@ func buildStddevSamp(aggFuncDesc *aggregation.AggFuncDesc, ordinal int) AggFunc
 	}
 }
 
+// buildJSONArrayagg builds the AggFunc implementation for function "json_arrayagg".
+func buildJSONArrayagg(aggFuncDesc *aggregation.AggFuncDesc, ordinal int) AggFunc {
+	base := baseJSONArrayaggFunc{
```
Comment (blacktear23) [inline executor/aggfuncs/builder.go:618]: Do we really need a `baseJSONArrayaggFunc`?

File: executor/aggfuncs/func_json_arrayagg.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+func (e *jsonArrayagg) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4JsonArrayagg)(pr)
+	if len(p.entries) == 0 {
+		chk.AppendNull(e.ordinal)
+		return nil
+	}
+
+	// appendBinary does not support some type such as uint8、types.time，so convert is needed here
+	for key, val := range p.entries {
```
Comment (blacktear23) [inline executor/aggfuncs/func_json_arrayagg.go:55]: rename `key` to `idx` should be better.

File: executor/aggfuncs/func_json_arrayagg.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+		realItem := item.Clone().GetValue()
+		switch x := realItem.(type) {
+		case nil, bool, int64, uint64, float64, string, json.BinaryJSON, *types.MyDecimal, []uint8, types.Time, types.Duration:
+			p.entries = append(p.entries, realItem)
+		default:
+			return 0, json.ErrUnsupportedSecondArgumentType.GenWithStackByArgs(x)
+		}
+	}
+	return 0, nil
```
Comment (blacktear23) [inline executor/aggfuncs/func_json_arrayagg.go:92]: please return correct `memDelta`

**Comment by arthuryangcs** [comment]
@blacktear23 PTAL

**Comment by blacktear23** [comment]
please add unit test for memDelta, you can reference from func_json_objectagg_test.go. Rest LGTM.

**Comment by blacktear23** [comment]
@XuHuaiyu PTAL

**Comment by arthuryangcs** [comment]
> please add unit test for memDelta, you can reference from func_json_objectagg_test.go. Rest LGTM.

@blacktear23  Done.

**Review by wshwsh12 — COMMENTED** [review]
Sorry for the late reply.   Thanks for your contribution.  
The logic for JSONArrayagg LGTM, but I have some problems about test code.

**Comment by arthuryangcs** [comment]
> Sorry for the late reply. Thanks for your contribution.
> The logic for JSONArrayagg LGTM, but I have some problems about test code.

@wshwsh12  Done.

**Comment by tisonkun** [comment]
@arthuryangcs you can cc @wshwsh12 to ping him. Most of our committer received many email every day and feel free to ping them when they don't repsonse.

/cc @wshwsh12

File: executor/aggfuncs/func_json_arrayagg_test.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+		// to adapt the `genSrcChk` Chunk format
+		entries = append(entries, nil)
+
+		aggTest := buildMultiArgsAggTester(ast.AggFuncJsonArrayagg, argTypes, mysql.TypeJSON, numRows, nil, json.CreateBinary(entries))
+
+		tests = append(tests, aggTest)
+	}
+
+	for _, test := range tests {
+		s.testMultiArgsAggFuncWithoutDistinct(c, test)
```
Comment (wshwsh12) [inline executor/aggfuncs/func_json_arrayagg_test.go:103]: json_arrayagg is a single arg aggfuncs, so I think we need use the test function for single args

File: executor/aggfuncs/func_json_arrayagg_test.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+		entries3 = append(entries3, entries1...)
+		entries3 = append(entries3, entries2...)
+
+		aggTest := buildMultiArgsAggTester(ast.AggFuncJsonArrayagg, argTypes, mysql.TypeJSON, numRows, json.CreateBinary(entries1), json.CreateBinary(entries2), json.CreateBinary(entries3))
+
+		tests = append(tests, aggTest)
+	}
+
+	for _, test := range tests {
+		s.testMultiArgsMergePartialResult(c, test)
```
Comment (wshwsh12) [inline executor/aggfuncs/func_json_arrayagg_test.go:69]: ditto

File: executor/aggfuncs/aggfunc_test.go [hunk 9]
```diff
@@ -806,6 +806,58 @@ func (s *testSuite) testMultiArgsAggFunc(c *C, p multiArgsAggTest) {
 	c.Assert(result, Equals, 0)
 }
 
+func (s *testSuite) testMultiArgsAggFuncWithoutDistinct(c *C, p multiArgsAggTest) {
```
Comment (wshwsh12) [inline executor/aggfuncs/aggfunc_test.go:809]: ditto

**Comment by arthuryangcs** [comment]
Done. 
/cc @wshwsh12

**Comment by arthuryangcs** [comment]
PTAL
/cc @wshwsh12

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**